x/programs/rust/examples/single_nft/src/lib.rs

```diff
@@ -0,0 +1,267 @@
+//! A basic NFT contract.
+//! The program serves as a non-fungible token with the ability to mint and burn.
+//! Only supports whole units with no decimal places.
+//!
+//! The NFT must support the common NFT metadata format.
+//! This includes the name, symbol, and URI of the NFT.
+use wasmlanche_sdk::{
+    memory::{Memory, Pointer},
+    program::Program,
+    public, state_keys,
+    types::Address,
+};
+
+pub mod example;
+
+/// The total supply of the NFT.
+/// # Safety
+/// In this program only one NFT of each type is intended to be minted.
+const TOTAL_SUPPLY: i64 = 1;
+
+/// The program storage keys.
+#[state_keys]
+enum StateKey {
+    /// The total supply of the token. Key prefix 0x0.
+    TotalSupply,
+    /// The name of the token. Key prefix 0x1.
+    Name,
+    /// The symbol of the token. Key prefix 0x2.
+    Symbol,
+    /// Metadata of the token. Key prefix 0x3.
+    Uri,
+    /// Balance of the NFT token by address. Key prefix 0x4(address).
+    Balance(Address),
+    /// Counter -- used to keep track of total NFTs minted. Key prefix 0x5.
+    Counter,
+    /// Owner -- used to keep track of the owner. Key prefix 0x6.
+    Owner,
+}
+
+/// Initializes the NFT with all required metadata.
+/// This includes the name, symbol, image URI, owner, and total supply.
+/// Returns true if the initialization was successful.
+#[public]
+pub fn init(
+    program: Program,
+    nft_name_ptr: i64,
+    nft_name_length: i64,
+    nft_symbol_ptr: i64,
+    nft_symbol_length: i64,
+    nft_uri_ptr: i64,
+    nft_uri_length: i64,
+) -> bool {
+    let nft_name_ptr = Memory::new(Pointer::from(nft_name_ptr));
+    let nft_name = unsafe { nft_name_ptr.range(nft_name_length as usize) };
```

Comment:
Just something to think about but perhaps we need a better abstraction :)

```
fn unsafe_something_cute(ptr: *const u8, len: u32) -> &[u8] {
    unsafe {
        let memory = Memory::new(Pointer::from(ptr));
        memory.range(len as usize)
    }
}
```

Reply:
I took a stab https://github.com/ava-labs/hypersdk/pull/585

Reply:
```
the trait bound `wasmlanche_sdk::memory::Pointer: std::convert::From<*const u8>` is not satisfied
the trait `std::convert::From<i64>` is implemented for `wasmlanche_sdk::memory::Pointer`
```

Reply:
I'm not sure why it's an issue since `From<*const u8>` is implemented for Pointer

Reply:
I think we can optimize this later since I couldn't get it to work. I agree the code is not pretty as could-be now though.